File: android/src/main/java/com/photovideoeditor/ui/components/EditorPanel.kt
package com.photovideoeditor.ui.components

import android.content.Context
import android.graphics.Color
import android.graphics.Typeface
import android.view.Gravity
import android.view.View
import android.widget.FrameLayout
import android.widget.HorizontalScrollView
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import com.photovideoeditor.ui.DesignTokens

/**
 * Shared building blocks for the editor's bottom panels (Crop / Adjust / Filters / Text / Stickers).
 *
 * Before this existed each panel hand-rolled its own header, chips and spacing, which is how the
 * Adjust panel ended up with four full-height rectangular buttons while Filters used thumbnail cards.
 * Every panel now composes the same three pieces — [editorPanelHeader], one slider, one
 * [editorToolStrip] — so heights, padding, radii, typography and selected states stay in step.
 *
 * Fixed vertical budget (see PANEL_* below): header 40dp + slider ~48dp + strip 76dp, so a panel
 * never grows past ~164dp and the photo keeps the rest of the screen.
 */
object EditorPanelMetrics {
  const val HEADER_HEIGHT_DP = 40
  const val SLIDER_HEIGHT_DP = 48
  const val STRIP_HEIGHT_DP = 76
  const val STRIP_ITEM_WIDTH_DP = 68
  const val ICON_SIZE_DP = 22
  const val THUMBNAIL_WIDTH_DP = 64
  const val THUMBNAIL_HEIGHT_DP = 72
  const val TOTAL_HEIGHT_DP = HEADER_HEIGHT_DP + SLIDER_HEIGHT_DP + STRIP_HEIGHT_DP
  /** Panel without the slider row (e.g. a picker that has nothing to scrub). */
  const val HEADER_AND_STRIP_HEIGHT_DP = HEADER_HEIGHT_DP + STRIP_HEIGHT_DP
}

/**
 * Compact panel header: `Title            [Reset]  Done`.
 *
 * "Done" is an action, not a tool — it gets a 44dp touch target via padding while staying visually
 * small, rather than becoming another full-width card.
 */
fun editorPanelHeader(
  context: Context,
  title: String,
  accentColor: Int,
  secondaryLabel: String? = null,
  onSecondary: (() -> Unit)? = null,
  onDone: () -> Unit
): LinearLayout = LinearLayout(context).apply {
  orientation = LinearLayout.HORIZONTAL
  gravity = Gravity.CENTER_VERTICAL
  setPadding(dpToPx(DesignTokens.spaceLg, context), 0, dpToPx(DesignTokens.spaceSm, context), 0)
  addView(TextView(context).apply {
    text = title
    textSize = 14f
    maxLines = 1
    includeFontPadding = false
    setTypeface(typeface, Typeface.BOLD)
    setTextColor(DesignTokens.textPrimary)
  }, LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, 1f))
  if (secondaryLabel != null && onSecondary != null) {
    addView(editorHeaderAction(context, secondaryLabel, DesignTokens.textSecondary, onSecondary))
  }
  addView(editorHeaderAction(context, "Done", accentColor, onDone))
}

/** Small text action sized for touch (44dp) without looking like a button. */
private fun editorHeaderAction(context: Context, label: String, color: Int, onClick: () -> Unit): TextView =
  TextView(context).apply {
    text = label
    textSize = 13f
    maxLines = 1
    gravity = Gravity.CENTER
    includeFontPadding = false
    setTypeface(typeface, Typeface.BOLD)
    setTextColor(color)
    minWidth = dpToPx(DesignTokens.touchTargetMin, context)
    minHeight = dpToPx(DesignTokens.touchTargetMin, context)
    val padH = dpToPx(DesignTokens.spaceMd, context)
    setPadding(padH, 0, padH, 0)
    isClickable = true
    isFocusable = true
    setOnClickListener { onClick() }
    applyPressScale()
  }

/** Horizontally scrollable strip of compact tool items; the container every panel's list sits in. */
fun editorToolStrip(context: Context, build: LinearLayout.() -> Unit): HorizontalScrollView =
  HorizontalScrollView(context).apply {
    isHorizontalScrollBarEnabled = false
    clipToPadding = false
    val padH = dpToPx(DesignTokens.spaceMd, context)
    setPadding(padH, 0, padH, 0)
    addView(LinearLayout(context).apply {
      orientation = LinearLayout.HORIZONTAL
      gravity = Gravity.CENTER_VERTICAL
      build()
    }, LinearLayout.LayoutParams(LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.MATCH_PARENT))
  }

/** An icon + single-line label entry in a strip, with an optional "value changed" dot. */
class EditorStripItem(
  val root: LinearLayout,
  private val iconHolder: FrameLayout,
  private val icon: ImageView,
  private val label: TextView,
  private val dot: View
) {
  /**
   * Selected state is a purple icon + label over a faint purple disc — never a filled purple tile,
   * which at this size reads as a giant button rather than a selection.
   */
  fun setSelected(selected: Boolean, accentColor: Int) {
    icon.setColorFilter(if (selected) accentColor else DesignTokens.textSecondary)
    label.setTextColor(if (selected) accentColor else DesignTokens.textSecondary)
    iconHolder.background = if (selected) {
      circleDrawable(DesignTokens.withAlphaPercent(accentColor, 18))
    } else {
      null
    }
    root.isSelected = selected
  }

  /** Tiny dot marking an adjustment that is no longer at its default value. */
  fun setModified(modified: Boolean, accentColor: Int) {
    dot.visibility = if (modified) View.VISIBLE else View.INVISIBLE
    dot.background = circleDrawable(accentColor)
  }
}

private fun circleDrawable(color: Int): android.graphics.drawable.GradientDrawable =
  android.graphics.drawable.GradientDrawable().apply {
    shape = android.graphics.drawable.GradientDrawable.OVAL
    setColor(color)
  }

fun editorStripItem(
  context: Context,
  iconRes: Int,
  labelText: String,
  onClick: () -> Unit
): EditorStripItem {
  val icon = ImageView(context).apply {
    setImageResource(iconRes)
    setColorFilter(DesignTokens.textSecondary)
  }
  val iconHolder = FrameLayout(context).apply {
    addView(
      icon,
      FrameLayout.LayoutParams(
        dpToPx(EditorPanelMetrics.ICON_SIZE_DP, context),
        dpToPx(EditorPanelMetrics.ICON_SIZE_DP, context),
        Gravity.CENTER
      )
    )
  }
  val label = TextView(context).apply {
    text = labelText
    textSize = 11f
    // Labels like "Brightness" and "Saturation" must never wrap to a second line.
    maxLines = 1
    includeFontPadding = false
    gravity = Gravity.CENTER
    setTextColor(DesignTokens.textSecondary)
    setPadding(0, dpToPx(DesignTokens.spaceXs, context), 0, 0)
  }
  val dot = View(context).apply { visibility = View.INVISIBLE }
  val root = LinearLayout(context).apply {
    orientation = LinearLayout.VERTICAL
    gravity = Gravity.CENTER_HORIZONTAL
    isClickable = true
    isFocusable = true
    contentDescription = labelText
    minimumHeight = dpToPx(DesignTokens.touchTargetMin, context)
    addView(
      iconHolder,
      LinearLayout.LayoutParams(dpToPx(32, context), dpToPx(32, context))
    )
    addView(label, LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT))
    addView(dot, LinearLayout.LayoutParams(dpToPx(4, context), dpToPx(4, context)).apply {
      topMargin = dpToPx(3, context)
    })
    setOnClickListener { onClick() }
    applyPressScale()
  }
  return EditorStripItem(root, iconHolder, icon, label, dot)
}

/**
 * Filter thumbnail: a preview of the current photo, outlined in purple when selected.
 * Deliberately a thin outline rather than a filled container, so the photo stays readable.
 */
class EditorFilterThumbnail(val root: LinearLayout, val image: ImageView, private val frame: FrameLayout, private val label: TextView) {
  fun setSelected(selected: Boolean, accentColor: Int) {
    frame.background = roundedDrawable(
      color = Color.TRANSPARENT,
      radiusDp = DesignTokens.radiusMd,
      context = root.context,
      strokeColor = if (selected) accentColor else Color.TRANSPARENT,
      strokeWidthDp = 2
    )
    label.setTextColor(if (selected) accentColor else DesignTokens.textSecondary)
    root.isSelected = selected
  }
}

fun editorFilterThumbnail(context: Context, labelText: String, onClick: () -> Unit): EditorFilterThumbnail {
  val image = ImageView(context).apply {
    scaleType = ImageView.ScaleType.CENTER_CROP
    background = roundedDrawable(DesignTokens.surfaceContainerHigh, DesignTokens.radiusMd, context)
    clipToOutline = true
    outlineProvider = android.view.ViewOutlineProvider.BACKGROUND
  }
  // The outline lives on a frame *around* the image so the stroke never crops the preview.
  val frame = FrameLayout(context).apply {
    val inset = dpToPx(2, context)
    setPadding(inset, inset, inset, inset)
    addView(
      image,
      FrameLayout.LayoutParams(
        dpToPx(EditorPanelMetrics.THUMBNAIL_WIDTH_DP, context),
        dpToPx(EditorPanelMetrics.THUMBNAIL_HEIGHT_DP, context)
      )
    )
  }
  val label = TextView(context).apply {
    text = labelText
    textSize = 11f
    maxLines = 1
    includeFontPadding = false
    gravity = Gravity.CENTER
    setTextColor(DesignTokens.textSecondary)
    setPadding(0, dpToPx(DesignTokens.spaceXs, context), 0, 0)
  }
  val root = LinearLayout(context).apply {
    orientation = LinearLayout.VERTICAL
    gravity = Gravity.CENTER_HORIZONTAL
    isClickable = true
    isFocusable = true
    contentDescription = labelText
    addView(frame, LinearLayout.LayoutParams(LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT))
    addView(label, LinearLayout.LayoutParams(dpToPx(EditorPanelMetrics.THUMBNAIL_WIDTH_DP + 8, context), LinearLayout.LayoutParams.WRAP_CONTENT))
    setOnClickListener { onClick() }
    applyPressScale()
  }
  return EditorFilterThumbnail(root, image, frame, label)
}

File: ios/Editor/UI/EditorPanel.swift
import UIKit

/// Shared building blocks for the editor's bottom panels (Crop / Adjust / Filters / Text / Stickers).
/// Mirrors `EditorPanel.kt` — keep the two in step so the platforms don't drift apart again.
///
/// Before this existed, Adjust and Filters shared one bar of plain text buttons. Every panel now
/// composes the same three pieces — `editorPanelHeader`, one slider, one strip — so heights,
/// padding, radii, typography and selected states stay consistent.
enum EditorPanelMetrics {
  static let headerHeight: CGFloat = 40
  static let sliderHeight: CGFloat = 48
  static let stripHeight: CGFloat = 76
  static let stripItemWidth: CGFloat = 58
  static let iconSize: CGFloat = 20
  static let thumbnailWidth: CGFloat = 64
  static let thumbnailHeight: CGFloat = 72
  static var totalHeight: CGFloat { headerHeight + sliderHeight + stripHeight }
}

/// Compact panel header: `Title            [Reset]  Done`.
/// "Done" is an action, not a tool — it gets a 44pt touch target while staying visually small.
func editorPanelHeader(
  title: String,
  accentColor: UIColor,
  secondaryLabel: String? = nil,
  onSecondary: (() -> Void)? = nil,
  onDone: @escaping () -> Void
) -> UIView {
  let container = UIStackView()
  container.axis = .horizontal
  container.alignment = .center
  container.distribution = .equalSpacing
  container.isLayoutMarginsRelativeArrangement = true
  container.layoutMargins = UIEdgeInsets(
    top: 0,
    left: DesignTokens.spaceMd,
    bottom: 0,
    right: DesignTokens.spaceMd
  )

  let titleLabel = UILabel()
  titleLabel.text = title
  titleLabel.font = .systemFont(ofSize: 15, weight: .semibold)
  titleLabel.textColor = DesignTokens.textPrimary
  container.addArrangedSubview(titleLabel)

  let actions = UIStackView()
  actions.axis = .horizontal
  actions.spacing = DesignTokens.spaceMd
  actions.alignment = .center

  if let secondaryLabel, let onSecondary {
    let reset = UIButton(type: .system)
    reset.setTitle(secondaryLabel, for: .normal)
    reset.setTitleColor(DesignTokens.textSecondary, for: .normal)
    reset.titleLabel?.font = .systemFont(ofSize: 13, weight: .regular)
    reset.addAction(UIAction { _ in onSecondary() }, for: .touchUpInside)
    reset.applyPressScale()
    actions.addArrangedSubview(reset)
  }

  let done = editorPanelDoneButton(accentColor: accentColor, action: onDone)
  actions.addArrangedSubview(done)

  container.addArrangedSubview(actions)
  return container
}

private func editorPanelDoneButton(accentColor: UIColor, action: @escaping () -> Void) -> UIButton {
  let control = UIButton(type: .system)
  var config = UIButton.Configuration.plain()
  config.title = "Done"
  config.baseForegroundColor = accentColor
  config.contentInsets = NSDirectionalEdgeInsets(top: 8, leading: 12, bottom: 8, trailing: 12)
  config.titleTextAttributesTransformer = UIConfigurationTextAttributesTransformer { incoming in
    var value = incoming; value.font = .systemFont(ofSize: 13, weight: .bold); return value
  }
  control.configuration = config
  control.widthAnchor.constraint(greaterThanOrEqualToConstant: DesignTokens.touchTargetMin).isActive = true
  control.heightAnchor.constraint(equalToConstant: DesignTokens.touchTargetMin).isActive = true
  control.addAction(UIAction { _ in action() }, for: .touchUpInside)
  control.applyPressScale()
  return control
}

/// Horizontally scrollable strip; the container every panel's list sits in.
func editorToolStrip(_ content: UIStackView) -> UIScrollView {
  let scroll = UIScrollView()
  scroll.showsHorizontalScrollIndicator = false
  scroll.alwaysBounceHorizontal = true
  scroll.delaysContentTouches = true
  scroll.canCancelContentTouches = true
  content.axis = .horizontal
  content.alignment = .fill
  scroll.addSubview(content)
  content.translatesAutoresizingMaskIntoConstraints = false
  NSLayoutConstraint.activate([
    content.leadingAnchor.constraint(equalTo: scroll.contentLayoutGuide.leadingAnchor, constant: DesignTokens.spaceMd),
    scroll.contentLayoutGuide.trailingAnchor.constraint(equalTo: content.trailingAnchor, constant: DesignTokens.spaceMd),
    content.topAnchor.constraint(equalTo: scroll.contentLayoutGuide.topAnchor),
    scroll.contentLayoutGuide.bottomAnchor.constraint(equalTo: content.bottomAnchor),
    content.heightAnchor.constraint(equalTo: scroll.frameLayoutGuide.heightAnchor),
  ])
  return scroll
}

/// Interactive icon + label button in an adjustment strip, with native press feedback.
final class EditorStripItemControl: UIButton {
  let iconHolder = UIView()
  let icon = UIImageView()
  let label = UILabel()
  let dot = UIView()
  private var action: (() -> Void)?

  init(systemName: String, labelText: String, action: @escaping () -> Void) {
    self.action = action
    super.init(frame: .zero)
    setupViews(systemName: systemName, labelText: labelText)
  }

  required init?(coder: NSCoder) { fatalError("init(coder:) has not been implemented") }

  private func setupViews(systemName: String, labelText: String) {
    isAccessibilityElement = true
    accessibilityLabel = labelText
    accessibilityTraits = .button

    icon.image = UIImage(systemName: systemName, withConfiguration: UIImage.SymbolConfiguration(pointSize: 16, weight: .regular))
    icon.tintColor = DesignTokens.textSecondary
    icon.contentMode = .scaleAspectFit
    icon.isUserInteractionEnabled = false

    iconHolder.layer.cornerRadius = 15
    iconHolder.isUserInteractionEnabled = false
    iconHolder.addSubview(icon)
    icon.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      iconHolder.widthAnchor.constraint(equalToConstant: 30),
      iconHolder.heightAnchor.constraint(equalToConstant: 30),
      icon.centerXAnchor.constraint(equalTo: iconHolder.centerXAnchor),
      icon.centerYAnchor.constraint(equalTo: iconHolder.centerYAnchor),
      icon.widthAnchor.constraint(equalToConstant: EditorPanelMetrics.iconSize),
      icon.heightAnchor.constraint(equalToConstant: EditorPanelMetrics.iconSize),
    ])

    label.text = labelText
    label.font = .systemFont(ofSize: 11, weight: .medium)
    label.textColor = DesignTokens.textSecondary
    label.textAlignment = .center
    label.numberOfLines = 1
    label.adjustsFontSizeToFitWidth = true
    label.minimumScaleFactor = 0.85
    label.isUserInteractionEnabled = false

    dot.layer.cornerRadius = 2
    dot.isHidden = true
    dot.isUserInteractionEnabled = false
    dot.widthAnchor.constraint(equalToConstant: 4).isActive = true
    dot.heightAnchor.constraint(equalToConstant: 4).isActive = true

    let column = UIStackView(arrangedSubviews: [iconHolder, label, dot])
    column.axis = .vertical
    column.alignment = .center
    column.spacing = DesignTokens.spaceXs
    column.isUserInteractionEnabled = false

    addSubview(column)
    column.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      widthAnchor.constraint(equalToConstant: EditorPanelMetrics.stripItemWidth),
      column.centerXAnchor.constraint(equalTo: centerXAnchor),
      column.centerYAnchor.constraint(equalTo: centerYAnchor),
      column.leadingAnchor.constraint(greaterThanOrEqualTo: leadingAnchor),
      column.trailingAnchor.constraint(lessThanOrEqualTo: trailingAnchor),
    ])

    addAction(UIAction { [weak self] _ in self?.action?() }, for: .touchUpInside)
    applyPressScale()
  }

  override func point(inside point: CGPoint, with event: UIEvent?) -> Bool {
    return bounds.insetBy(dx: -8, dy: -8).contains(point)
  }

  func setSelected(_ selected: Bool, accentColor: UIColor) {
    icon.tintColor = selected ? accentColor : DesignTokens.textSecondary
    label.textColor = selected ? accentColor : DesignTokens.textSecondary
    iconHolder.backgroundColor = selected ? accentColor.withAlphaComponent(0.18) : .clear
    accessibilityTraits = selected ? [.button, .selected] : .button
  }

  func setModified(_ modified: Bool, accentColor: UIColor) {
    dot.isHidden = !modified
    dot.backgroundColor = accentColor
  }
}

/// An icon + single-line label entry in a strip, with an optional "value changed" dot.
final class EditorStripItem {
  let control: EditorStripItemControl
  var root: UIView { control }

  init(control: EditorStripItemControl) {
    self.control = control
  }

  func setSelected(_ selected: Bool, accentColor: UIColor) {
    control.setSelected(selected, accentColor: accentColor)
  }

  func setModified(_ modified: Bool, accentColor: UIColor) {
    control.setModified(modified, accentColor: accentColor)
  }
}

func editorStripItem(systemName: String, labelText: String, action: @escaping () -> Void) -> EditorStripItem {
  let control = EditorStripItemControl(systemName: systemName, labelText: labelText, action: action)
  return EditorStripItem(control: control)
}

/// Interactive filter thumbnail button with native touch actions and selection border.
final class EditorFilterThumbnailControl: UIButton {
  let image = UIImageView()
  let frameView = UIView()
  let label = UILabel()
  private var action: (() -> Void)?

  init(labelText: String, action: @escaping () -> Void) {
    self.action = action
    super.init(frame: .zero)
    setupViews(labelText: labelText)
  }

  required init?(coder: NSCoder) { fatalError("init(coder:) has not been implemented") }

  private func setupViews(labelText: String) {
    isAccessibilityElement = true
    accessibilityLabel = labelText
    accessibilityTraits = .button

    image.contentMode = .scaleAspectFill
    image.clipsToBounds = true
    image.layer.cornerRadius = DesignTokens.radiusMd
    image.backgroundColor = DesignTokens.surfaceContainerHigh
    image.isUserInteractionEnabled = false

    frameView.layer.cornerRadius = DesignTokens.radiusMd + 2
    frameView.layer.borderWidth = 2
    frameView.layer.borderColor = UIColor.clear.cgColor
    frameView.isUserInteractionEnabled = false

    frameView.addSubview(image)
    image.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      image.widthAnchor.constraint(equalToConstant: EditorPanelMetrics.thumbnailWidth),
      image.heightAnchor.constraint(equalToConstant: EditorPanelMetrics.thumbnailHeight),
      image.leadingAnchor.constraint(equalTo: frameView.leadingAnchor, constant: 2),
      image.trailingAnchor.constraint(equalTo: frameView.trailingAnchor, constant: -2),
      image.topAnchor.constraint(equalTo: frameView.topAnchor, constant: 2),
      image.bottomAnchor.constraint(equalTo: frameView.bottomAnchor, constant: -2),
    ])

    label.text = labelText
    label.font = .systemFont(ofSize: 11, weight: .medium)
    label.textColor = DesignTokens.textSecondary
    label.textAlignment = .center
    label.numberOfLines = 1
    label.adjustsFontSizeToFitWidth = true
    label.minimumScaleFactor = 0.8
    label.isUserInteractionEnabled = false

    let column = UIStackView(arrangedSubviews: [frameView, label])
    column.axis = .vertical
    column.alignment = .center
    column.spacing = DesignTokens.spaceXs
    column.isUserInteractionEnabled = false

    addSubview(column)
    column.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      widthAnchor.constraint(equalToConstant: EditorPanelMetrics.thumbnailWidth + 12),
      column.centerXAnchor.constraint(equalTo: centerXAnchor),
      column.centerYAnchor.constraint(equalTo: centerYAnchor),
      label.widthAnchor.constraint(equalTo: widthAnchor),
    ])

    addAction(UIAction { [weak self] _ in self?.action?() }, for: .touchUpInside)
    applyPressScale()
  }

  override func point(inside point: CGPoint, with event: UIEvent?) -> Bool {
    return bounds.insetBy(dx: -8, dy: -8).contains(point)
  }

  func setSelected(_ selected: Bool, accentColor: UIColor) {
    frameView.layer.borderColor = (selected ? accentColor : UIColor.clear).cgColor
    frameView.layer.borderWidth = 2
    label.textColor = selected ? accentColor : DesignTokens.textSecondary
    accessibilityTraits = selected ? [.button, .selected] : .button
  }
}

/// Filter thumbnail: a preview of the current photo, outlined in purple when selected.
final class EditorFilterThumbnail {
  let control: EditorFilterThumbnailControl
  var root: UIView { control }
  var image: UIImageView { control.image }

  init(control: EditorFilterThumbnailControl) {
    self.control = control
  }

  func setSelected(_ selected: Bool, accentColor: UIColor) {
    control.setSelected(selected, accentColor: accentColor)
  }
}

func editorFilterThumbnail(labelText: String, action: @escaping () -> Void) -> EditorFilterThumbnail {
  let control = EditorFilterThumbnailControl(labelText: labelText, action: action)
  return EditorFilterThumbnail(control: control)
}

/// The editor's one slider: a `Label ......... value` line above a thin track, so a panel never
/// needs a separate caption row and the same instance can be re-pointed at whichever adjustment is
/// selected. Mirrors `EditorToolSlider.kt`.
final class EditorValueSlider: UIView {
  let slider = UISlider()
  private let titleLabel = UILabel()
  private let valueLabel = UILabel()
  private let neutralTick = UIView()

  /// Value at which the underlying adjustment is neutral, for bipolar ranges. When set, a tick is
  /// drawn there so it is obvious where "no change" sits.
  var neutralValue: Float? {
    didSet { updateNeutralTick() }
  }

  var titleText: String? {
    get { titleLabel.text }
    set { titleLabel.text = newValue }
  }

  var valueText: String? {
    get { valueLabel.text }
    set { valueLabel.text = newValue }
  }

  init(accentColor: UIColor) {
    super.init(frame: .zero)
    titleLabel.font = .systemFont(ofSize: 12, weight: .medium)
    titleLabel.textColor = DesignTokens.textSecondary
    valueLabel.font = .monospacedDigitSystemFont(ofSize: 12, weight: .medium)
    valueLabel.textColor = DesignTokens.textSecondary
    valueLabel.textAlignment = .right

    slider.minimumTrackTintColor = accentColor
    slider.maximumTrackTintColor = UIColor.white.withAlphaComponent(0.28)
    if let thumb = UIImage(systemName: "circle.fill")?.withConfiguration(UIImage.SymbolConfiguration(pointSize: 12)) {
      slider.setThumbImage(thumb.withTintColor(.white, renderingMode: .alwaysOriginal), for: .normal)
    }

    neutralTick.backgroundColor = DesignTokens.textSubtle
    neutralTick.isHidden = true
    // Sits above the track (added last) but must never swallow a drag toward it.
    neutralTick.isUserInteractionEnabled = false

    let caption = UIStackView(arrangedSubviews: [titleLabel, valueLabel])
    caption.axis = .horizontal
    caption.distribution = .fill
    valueLabel.setContentHuggingPriority(.required, for: .horizontal)

    let column = UIStackView(arrangedSubviews: [caption, slider])
    column.axis = .vertical
    column.spacing = 2
    column.isLayoutMarginsRelativeArrangement = true
    column.layoutMargins = UIEdgeInsets(top: 4, left: DesignTokens.spaceLg, bottom: 4, right: DesignTokens.spaceLg)

    addSubview(column)
    addSubview(neutralTick)
    column.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      column.leadingAnchor.constraint(equalTo: leadingAnchor),
      column.trailingAnchor.constraint(equalTo: trailingAnchor),
      column.topAnchor.constraint(equalTo: topAnchor),
      column.bottomAnchor.constraint(equalTo: bottomAnchor),
    ])
  }

  required init?(coder: NSCoder) { fatalError("init(coder:) is not used") }

  override func layoutSubviews() {
    super.layoutSubviews()
    updateNeutralTick()
  }

  private func updateNeutralTick() {
    guard let neutralValue, slider.maximumValue > slider.minimumValue, slider.bounds.width > 0 else {
      neutralTick.isHidden = true
      return
    }
    neutralTick.isHidden = false
    let fraction = CGFloat((neutralValue - slider.minimumValue) / (slider.maximumValue - slider.minimumValue))
    let trackFrame = slider.convert(slider.bounds, to: self)
    let x = trackFrame.minX + trackFrame.width * fraction
    neutralTick.frame = CGRect(x: x - 0.5, y: trackFrame.midY - 5, width: 1, height: 10)
  }
}
